### ISSUE TITLE
XSPEC fits failing because response files can't be read in

### ISSUE THREAD

**Issue by DavidT3** [DavidT3/XGA#745]
This is happening on my new install on MacOS 12.1 on my new ARM M1 macbook pro - SAS v20.0.0, XSPEC v12.12.0.

Not sure what the cause of this is yet, but on this setup XSPEC fits are failing because it can't read in the response files specified in the spectrum headers. It seems to be happening because when XSPEC reads it in it can't deal with & at the end of the filename.

I don't _think_ its XSPEC causing the issue here - having looked at spectrum headers for RESPFILE, ANCRFILE for files generated on the mac and another computer with a known setup (SAS v17.0.0, XSPEC v12.10.1) there is a difference in the headers.

Spectra generated on the laptop have an extra CONTINUE for the response file and ancillary file entries, but its empty (just contains a ''). As & is used to denote a break in a header entry that leads into a continue, there is one at the end of the actual entry that leads into the empty CONTINUE.

Spectra generated on the known good computer do not have this.

Don't know whether this is because of the difference in SAS version or something else, but I am going looking.

**Comment by DavidT3** [comment]
I thought perhaps it was a fitsio version thing, so I reverted from fitsio 1.1.6 to 1.1.4 on the laptop and the same problem occured again

**Comment by DavidT3** [comment]
Accessing the ANCRFILE from spectrum.header (for instance) actually provides the same result - a string with a & at the end (even though the arf path stored in the spectrum arf attribute is just the normal path with no &). This makes me wonder if there is a problem with cfitsio library? The odd thing though is that its being written like that at all

**Comment by DavidT3** [comment]
Addressing this issue again because it has now started happening on Apollo as well, the headers are being written with an & on the end, presumably a bug due to them needing continuing (which has &s on the end of each continued line).

Taking another look at the _update_spec_headers method of the Spectrum class (which is presumably where the issue lies), I see that I'm actually writing with the astropy implementation - I seem to remember FITSIO behaved weirdly and I didn't like it here (makes you wonder if that strange behaviour that caused me to use fitsio in the first place is now fixed).

So perhaps its an astropy versioning issue

**Comment by DavidT3** [comment]
So its definitely not that XSPEC has changed (not that I ever really thought it was), I've checked some past generated spectra and they **don't** have & on the end of them, so it's not that XSPEC can no longer handle the &s, it's the writing process.

Logically that would likely mean the astropy version, but I wonder if something in the backend has changed a library or something. I've asked Jessica which versions of software she used for the tutorials, and whether any issues fitting spectra occurred. 

I can also ask Paul to ask his students, as clearly they aren't having problems using XGA - I know Grace has been doing a bunch of XSPEC fits at least.

**Comment by DavidT3** [comment]
This appears to be happening on my ICER install as well. With SAS 20.0.0 and HEASoft v6.31.1.

Astropy version is 5.1.1 - all I specify in the requirements file is Astropy > 4 - so I really need someone with a working install to tell me what their astropy version is.

**Comment by DavidT3** [comment]
I have matched astropy versions with someone who has a working install, Astropy 5.1.1 - this problem occurs.

Okay so when I switch back to SAS v18.0.0 from SAS v20.0.0 on Apollo this problem no longer occurs (why is it _**ALWAYS**_ SAS...)

I have two working theories - one is driven by my noticing that the pre-spectrum head updating by XGA files from SAS v18 don't actually have entries for RESP_FILE, ANCR_FILE, or BACK_FILE at all. I'm 90% sure that the SAS v20 ones do have entries (they're just None or something). I will double check this. Anyway, this makes me wonder if Astropy behaves differently updating a header than writing one from scratch. 

The second theory would be more annoying, that SAS has got a fits library that is taking over from whatever Astropy uses in the backend. I hope this isn't the cause because then I don't know how I might fix it.

**Comment by DavidT3** [comment]
So I am correct in that the SAS v20 spectra (prior to XGA Spectrum objects adding headers for the arf, rmf, and backspec) do have entries for RESPFILE, ANCRFILE, and BACKFILE. They are all set to 'NONE    '

And what seems to be happening after astropy is used to change the value of those headers is that the new paths are written with CONTINUE (as is necessary as they are long), but they all have a final CONTINUE which is just '' and as such is unnecessary - hence why the path read by XSPEC has an & on the end.

**Comment by DavidT3** [comment]
It would appear that the first thing I tried has fixed it - though haven't got to the spectrum running stage yet.

I decided to try deleting the headers before writing them again (just with del spec_fits["SPECTRUM"].header["RESPFILE"] for instance). Looking at the spectrum file it seems like it has written the headers properly, without an extra continue on the end. I'll make sure it gets all the way through to fitting though.

Also I will need to account for the fact that some versions of SAS won't make those headers initially - so will have to check for their presence before deleting them.
